src/main/kotlin/org/rust/ide/refactoring/move/RsMoveFilesOrDirectoriesProcessor.kt

```diff
@@ -0,0 +1,304 @@
+ * - Move mod-declaration to new parent mod
+ * - Update necessary imports in other files
+ * - Update necessary paths in other files (some usages could still remain invalid because of glob-imports)
+ *     We replace path with absolute if there are few usages with same path in file, otherwise add new import
+ * - Update relative paths in moved file
+ */
+class RsMoveFilesOrDirectoriesProcessor(
+    private val project: Project,
+    private val elementsToMove: Array<PsiElement>,
+    private val newParent: PsiDirectory,
+    private val newParentMod: RsMod,
+    moveCallback: MoveCallback?,
+    doneCallback: Runnable
+) : MoveFilesOrDirectoriesProcessor(
+    project,
+    elementsToMove,
+    newParent,
+    true,
+    true,
+    true,
+    moveCallback,
+    doneCallback
+) {
+
+    private val psiFactory = RsPsiFactory(project)
+    private val movedFile: RsFile = elementsToMove[0] as RsFile
+
+    // keys --- `RsPath`s inside movedFile
+    // outsideReferencesMap[path] --- target element for path reference
+    private lateinit var outsideReferencesMap: Map<RsPath, RsElement>
+    private lateinit var conflictsDetector: RsMoveFilesOrDirectoriesConflictsDetector
+
+    override fun doRun() {
+        checkMove()
+        super.doRun()
+    }
+
+    private fun checkMove() {
+        // TODO: support move multiply files
+        check(elementsToMove.size == 1)
+
+        check(newParentMod.crateRoot == movedFile.crateRoot)
+        movedFile.modName?.let {
+            if (newParentMod.getChildModule(it) != null) {
+                throw IncorrectOperationException("Cannot move. Mod with same crate relative path already exists")
+            }
+        }
+    }
+
+    override fun preprocessUsages(refUsages: Ref<Array<UsageInfo>>): Boolean {
+        val usages = refUsages.get()
+        val conflicts = MultiMap<PsiElement, String>()
+
+        val success = ProgressManager.getInstance().runProcessWithProgressSynchronously(
+            { runReadAction { detectVisibilityProblems(usages, conflicts) } },
+            RefactoringBundle.message("detecting.possible.conflicts"),
+            true,
+            project
+        )
+        if (!success) return false
+
+        return showConflicts(conflicts, usages)
+    }
+
+    private fun detectVisibilityProblems(usages: Array<UsageInfo>, conflicts: MultiMap<PsiElement, String>) {
+        outsideReferencesMap = collectOutsideReferencesFromMovedFile()
+        conflictsDetector = RsMoveFilesOrDirectoriesConflictsDetector(movedFile, newParentMod, outsideReferencesMap)
+        conflictsDetector.detectVisibilityProblems(usages, conflicts)
+    }
+
+    private fun collectOutsideReferencesFromMovedFile(): MutableMap<RsPath, RsElement> {
+        val paths = movedFile.descendantsOfType<RsPath>()
+        val outsideReferencesMap = mutableMapOf<RsPath, RsElement>()
+        for (path in paths) {
+            val target = path.reference?.resolve() ?: continue
```

Comment:
What about multi-resolve cases?

Reply:
Good question, I checked usages of `multiResolve` and [`RsMultiResolveTest`](https://github.com/intellij-rust/intellij-rust/blob/b722b1f00112f9307ab46f19da23e6f1092f1fcf/src/test/kotlin/org/rust/lang/core/resolve/RsMultiResolveTest.kt), and looks like cases when multi-resolve is needed are rather specific: incorrect code (e.g. duplicate methods/fields) or multiple items with same name in mod (e.g. mod has function and child mod which has same name). So I think we can ignore multi-resolve for now, and add support for it later